### PR TITLE
[FOCAL-97, FOCAL-98] Construction of the FOCAL-H geometry and Hit simulation

### PR DESCRIPTION
**Pull request by hahassan7**
The HCAL geometry has been implemented into the Detectot.h/.cxx files along with the Hit management in the HCAL sensitive volume.

Things that still needs to be worked on:
1. The way the geometry file is loaded is now hard coded in the build_geometry.C file and there's no way to loaded it from the outside. So we need to make a config param that takes the geometry file with making a default if the config param was not called.
2. The addition of hit's energy. Currently for the Sandwich calorimeter the hit energy is added to the parent if the hit is in the same cell (volume) as the parent (same as AliRoot). But for the Spaghetti colorimeter, the hits are added if it is in the same cell as the parent, while for AliRoot the hit's energy is added if it falls in the same scintillator fiber.

### CODE REVIEW

File: run/o2simdefaultdetectorlist.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "MCH",
     "MID",
     "ZDC",
+    "FOC",
```
Comment (f3sch) [inline run/o2simdefaultdetectorlist.json]: FOCal is not part of ALICE2, no?

Comment (sawenzel) [reply run/o2simdefaultdetectorlist.json]: I agree. FOCAL should probably be in the ALICE2.1 detector list.

Comment (f3sch) [reply run/o2simdefaultdetectorlist.json]: Already is :)

File: run/O2HitMerger.h [hunk 2]
```diff
@@ -982,6 +983,10 @@ void O2HitMerger::initDetInstances()
       mDetectorInstances[i] = std::move(std::make_unique<o2::zdc::Detector>(true));
       counter++;
     }
+    if (i == DetID::FOC) {
+      mDetectorInstances[i] = std::move(std::make_unique<o2::focal::Detector>(true, gSystem->ExpandPathName("$VMCWORKDIR/Detectors/Geometry/FOC/geometryFiles/geometry_Spaghetti.txt")));
```
Comment (sawenzel) [inline run/O2HitMerger.h]: Do we need to use `$VMCWORKDIR` ... or above `$GEOMDIR`? Best would be to use `$O2_ROOT` which is well defined and used elsewhere. (I am aware that VMCWORKDIR is used in some other places but I think `O2_ROOT` is better).

Comment (sawenzel) [reply run/O2HitMerger.h]: Regard this comment as a minor wish. We can also leave as is (I realized that VMCWORKDIR + GEOMPATH are in use ... although they are a bit historic)

Comment (mfasDa) [reply run/O2HitMerger.h]: I agree it would be better to use `O2_ROOT` instead.

File: Detectors/FOCAL/simulation/src/Detector.cxx [hunk 3]
```diff
@@ -11,30 +11,83 @@
...
+
+  mMedSensHCal = getMediumID(ID_SC);
 }
 
 Bool_t Detector::ProcessHits(FairVolume* v)
 {
-  int track = fMC->GetStack()->GetCurrentTrackNumber(),
-      directparent = fMC->GetStack()->GetCurrentParentTrackNumber();
+  int track = gMC->GetStack()->GetCurrentTrackNumber(),
+      directparent = gMC->GetStack()->GetCurrentParentTrackNumber();
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/src/Detector.cxx]: I would avoid using the gMC variable and stick to the VMC app provided by FairModule.

File: Detectors/FOCAL/simulation/src/Detector.cxx [hunk 4]
```diff
@@ -58,14 +111,88 @@ Bool_t Detector::ProcessHits(FairVolume* v)
...
-      mCurrentSuperparent = AddSuperparent(track, fMC->TrackPid(), fMC->Etot());
+      mCurrentSuperparent = AddSuperparent(track, gMC->TrackPid(), gMC->Etot());
       mCurrentParentID = track;
     }
     mCurrentTrack = track;
   }
+
+  // Processing HCAL hits
+  if (gMC->CurrentMedium() == mMedSensHCal) {
+    LOG(debug) << "We are in sensitive volume " << v->GetName() << ": " << gMC->CurrentVolPath() << std::endl;
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/src/Detector.cxx]: The endl is not really needed.

File: Detectors/FOCAL/simulation/src/Detector.cxx [hunk 5]
```diff
@@ -261,14 +395,347 @@ AliMixture( 8, "Ceramic    $", acer, zcer, denscer, 2, wcer);
...
+  // Alignable volumes are:
+
+  // AddAlignableVolumesECAL();
+  addAlignableVolumesHCAL();
+}
+
+//____________________________________________________________________________
+void Detector::addAlignableVolumesHCAL() const
+{
+  TString vpsector = "/cave_1/barrel_1/FOCAL_1/HCAL_1";
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/src/Detector.cxx]: FOCAL is a forward detector, looks strange to me to put it into the volume `barrel_1`. I think FOCAL_1 should be a separate volume directly under cave.

File: Detectors/FOCAL/simulation/src/Detector.cxx [hunk 6]
```diff
@@ -261,14 +395,347 @@ AliMixture( 8, "Ceramic    $", acer, zcer, denscer, 2, wcer);
...
+
+      // Adding the Tower to the HCAL
+      volHCAL->AddNode(volTowerHCAL, NumTowers, trans);
+
+      NumTowers++;
+    }
+  }
+
+  std::cout << "Number of Towers is: " << (NumTowers - 1) << std::endl;
+  std::cout << "Number of tubes is: " << (NumTubes - 1) * (NumTowers - 1) << std::endl;
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/src/Detector.cxx]: Use `LOG(...)` instead

File: Detectors/FOCAL/simulation/src/Detector.cxx [hunk 7]
```diff
@@ -261,14 +395,347 @@ AliMixture( 8, "Ceramic    $", acer, zcer, denscer, 2, wcer);
...
+        continue;
+      }
+
+      // Adding the Tower to the HCAL
+      volHCAL->AddNode(volTower, NumTowers, trans);
+
+      NumTowers++;
+    }
+  }
+  std::cout << "Number of Towers is: " << (NumTowers - 1) << std::endl;
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/src/Detector.cxx]: Another `LOG(...)`

File: Detectors/FOCAL/simulation/data/simcuts.dat [hunk 8]
```diff
@@ -0,0 +1,16 @@
+* FOCAL
```
Comment (mfasDa) [inline Detectors/FOCAL/simulation/data/simcuts.dat]: @sawenzel How do we treat detector-specific simcuts? Can they stay in the FOCAL directory or should they be merged with the global simcuts?

**Review by mfasDa — CHANGES_REQUESTED** [review]

File: run/o2simdefaultdetectorlist.json (hunk 1 above)
Comment (hahassan7) [reply run/o2simdefaultdetectorlist.json]: Sorry I didn't see it in ALICE2.1.

File: run/O2HitMerger.h (hunk 2 above)
Comment (hahassan7) [reply run/O2HitMerger.h]: Done. Using O2_ROOT now.

File: Detectors/FOCAL/simulation/src/Detector.cxx (hunk 3 above)
Comment (hahassan7) [reply Detectors/FOCAL/simulation/src/Detector.cxx]: Using fMC now instead.

File: Detectors/FOCAL/simulation/src/Detector.cxx (hunk 4 above)
Comment (hahassan7) [reply Detectors/FOCAL/simulation/src/Detector.cxx]: Done. Removed.

File: Detectors/FOCAL/simulation/src/Detector.cxx (hunk 5 above)
Comment (hahassan7) [reply Detectors/FOCAL/simulation/src/Detector.cxx]: FOCAL has been moved to caveRB24.

File: Detectors/FOCAL/simulation/src/Detector.cxx (hunk 6 above)
Comment (hahassan7) [reply Detectors/FOCAL/simulation/src/Detector.cxx]: Done.

File: Detectors/FOCAL/simulation/src/Detector.cxx (hunk 7 above)
Comment (hahassan7) [reply Detectors/FOCAL/simulation/src/Detector.cxx]: Done.

File: run/o2simdefaultdetectorlist.json (hunk 1 above)
Comment (f3sch) [reply run/o2simdefaultdetectorlist.json]: np

**Review by mfasDa — DISMISSED** [review]
I would say it is ready to merge from the FOCAL side. @sawenzel Please have a final look!

**Comment by sawenzel** [comment]
fullCI failing with codechecker complaints:
```
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/13225-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:387:52: error: use nullptr [modernize-use-nullptr]
/sw/SOURCES/O2/13225-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:490:39: error: use nullptr [modernize-use-nullptr]
/sw/SOURCES/O2/13225-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:500:38: error: use nullptr [modernize-use-nullptr]
/sw/SOURCES/O2/13225-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:509:39: error: use nullptr [modernize-use-nullptr]
/sw/SOURCES/O2/13225-slc8_x86-64/0/Detectors/FOCAL/simulation/src/Detector.cxx:525:38: error: use nullptr [modernize-use-nullptr]
```

**Pull request merged**